sqlx-core/src/postgres/types/ltree.rs

```diff
@@ -0,0 +1,72 @@
+use std::fmt::{self, Display, Formatter};
+use std::str::FromStr;
+use crate::decode::Decode;
+use crate::encode::{Encode, IsNull};
+use crate::error::{Error, BoxDynError};
+use crate::postgres::{PgArgumentBuffer, PgValueFormat, PgTypeInfo, PgValueRef, Postgres};
+use crate::types::Type;
+
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct PgLTree {
+    labels: Vec<String>
+}
+
+impl PgLTree {
+    pub fn push(&mut self, label: String) -> Result<(), Error> {
+        if label.chars().all(|c| c.is_ascii_alphabetic() || c.is_ascii_digit() || c == '_') {
+            self.labels.push(label);
+            Ok(())
+        } else {
+            Err(Error::InvalidLtreeLabel)
+        }
+    }
+}
+
+impl FromStr for PgLTree {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Error> {
+        Ok(
+            Self {
+                labels: s.split('.').map(|s| s.to_owned()).collect()
+            }
+        )
+    }
+}
+
+impl Display for PgLTree {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.labels.join("."))
+    }
+}
+
+impl Type<Postgres> for PgLTree {
+    fn type_info() -> PgTypeInfo {
+        PgTypeInfo::LTREE
+    }
+}
+
+impl Encode<'_, Postgres> for PgLTree {
+    fn encode_by_ref(&self, buf: &mut PgArgumentBuffer) -> IsNull {
+        buf.extend(1i8.to_le_bytes());
+        buf.extend(self.to_string().as_bytes());
```

Comment:
Here I'd look into a way of having the Display implementation write directly into `buf`.

Reply:
It could be done by implementing something like this on PgArgumentBuffer

```rust
    pub(crate) fn add_display<T: fmt::Display>(&mut self, value: T) {
        struct VecFmt<'a>(&'a mut Vec<u8>);

        impl<'a> fmt::Write for VecFmt<'a> {
            fn write_char(&mut self, c: char) -> fmt::Result {
                self.write_str(c.encode_utf8(&mut [0u8; 4]))
            }

            fn write_str(&mut self, s: &str) -> fmt::Result {
                self.0.extend_from_slice(s.as_bytes());
                Ok(())
            }
        }

        write!(VecFmt(&mut self.buffer), "{}", value)
            .expect("Display implementation panicked while writing to PgArgumentBuffer");
    }
```

or duplicating the `Display` implementation inside `Encode` and avoiding the dynamic dispatch of the `Display` `Formatter`